pkg/ratelimit/rate_limit.go

```diff
@@ -0,0 +1,71 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ratelimit
+
+import (
+	"context"
+
+	"github.com/throttled/throttled"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+)
+
+// RateLimiter can be used to rate limit access to a resource (using a key string).
+type RateLimiter interface {
+	RateLimit(key string, class []string) (bool, *Result)
```

Comment:
Switched to `...string` and added missing documentation. The second value is rate limiting metadata, and are used to set response headers for HTTP and gRPC requests.

Also added a `ratelimit.Require` method for cases where we do not really care about such metadata (basically anything that does not involve a request context).

Should we specialize this into `ratelimit.RequireGatewayUp()`, `ratelimit.RequireApplicationDown()`, ....? I am a bit hesitant because that would basically recompute the rate limiting key for every iteration even when it is fixed.

I think it might be a cleaner solution to instead have `ratelimit.GatewayUpKey() string`, `ratelimit.GatewayUpClasses() ...string`, etc, that the caller can use to compute the keys once.

Reply:
> Should we specialize this into `ratelimit.RequireGatewayUp()`, `ratelimit.RequireApplicationDown()`, ....? I am a bit hesitant because that would basically recompute the rate limiting key for every iteration even when it is fixed.

No we shouldn't indeed